### DISCUSSION TITLE
Support for pipeline input.

### DISCUSSION THREAD

**Discussion by hsianktin** [TheR1D/shell_gpt#35] [Ideas]
First, thanks for the great implementation!
I noticed that in the [PowerShellAI](https://github.com/dfinke/PowerShellAI) module, there is a very interesting application of pipelining.
Although we can do similar things using `sgpt -se`, it will be nice to have similar functionality.

```
git status | ai "create a detailed git message"
```

```
Commit message:
Added PowerShellAI.psd1, README.md, changelog.md, and Public/ai.ps1 to dcf-spike-piping-to-ai-function branch. Updated PowerShellAI.psd1 and README.md with new changes. Added changelog.md to track changes. Added Public/ai.ps1 to enable piping to AI function.
```

Handling pipeline input in Python seems straightforward. I just asked ChatGPT about this :)
```Python
import sys

# Read input from pipeline
input_data = sys.stdin.read()
```

**Comment by TheR1D** [comment]
But isn't it possible to do something like: `"create a detailed git message $(git status)"` in PowerShell?

**Reply by hsianktin** [reply]
Yes. In general, one can always use the "$(function)" to do anything that can be done via pipelining.
I'm just hoping to add flexibility to the way of using the shell gpt program.
One possible benefit is that pipelining allows the shell to autocomplete better? 

On the other hand, the fish shell seems to be not quite compatible with this syntax.
So, allowing for pipelining may improve the compatibility by allowing for a way that is less shell-specific?
```
fish: $(...) is not supported. In fish, please use '(whoami)'.
sgpt -se "get the uid and gid for user $(whoami)"
```
The workaround is kind of ugly and strange:
```fish
sgpt -se "get the uid and gid for user "(whoami) 
```

In addition, there might also be possible use cases where multiple pipes are used, and shell gpt is the last pipe.
For those cases, I think allowing for pipeline input will be a good idea.

When I have time, I can see if I can implement this myself.

**Comment by lpurdy01** [comment]
I don't know why, but I often have `failed to execute` returns when I am running a command like $(cat file.txt) that I don't get if `cat` is first to run and then piped into sgpt. #93

**Comment by TheR1D** [comment]
Implemented https://github.com/TheR1D/shell_gpt/pull/163
